pandas/core/groupby/grouper.py

```diff
@@ -508,6 +508,15 @@ def _get_grouper(obj, key=None, axis=0, level=None, sort=True,
             warnings.warn(msg, FutureWarning, stacklevel=5)
             key = list(key)
 
+    if callable(key) or isinstance(key, dict):
```

Comment:
Is the `isinstance(key, dict)` new here? If so probably want a test case to cover

Reply:
(done - expanded the test)

Reply:
can you integrate this with the below (e.g. i/elif maybe enough)?
can add a top-levevel comment on what is going on here